docker/docker_client.go

```diff
@@ -446,8 +416,10 @@ func (c *dockerClient) ping() (*pingResponse, error) {
 		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
 			return nil, errors.Errorf("error pinging repository, response code %d", resp.StatusCode)
 		}
+		chs := parseAuthHeader(resp.Header)
+		c.saveChallenges(chs, resp.Request.URL)
```

Comment:
This is the _only_ `saveChallenges` call everywhere ever; if that’s all that is needed, we can go back to a single `dockerClient.wwwAuthenticate` and have the code noticeably simpler (no map, no URL normalization)